resource-aggregate/events/deviceMetadataSnapshotTaken.go

```diff
@@ -93,8 +93,22 @@ func (e *DeviceMetadataSnapshotTaken) HandleDeviceMetadataSnapshotTaken(ctx cont
 	return nil
 }
 
-func (e *DeviceMetadataSnapshotTaken) HandleDeviceMetadataUpdatePending(ctx context.Context, updatePending *DeviceMetadataUpdatePending) error {
+func (e *DeviceMetadataSnapshotTaken) HandleDeviceMetadataUpdatePending(ctx context.Context, updatePending *DeviceMetadataUpdatePending, isCmd bool) error {
+	now := time.Now()
+	if !updatePending.IsValid(now) {
+		if isCmd {
```

Comment:
It's always a command. If you do the validation at a proper place, you can remove this unreadable hack :)

Reply:
No. It's called during load events from eventstore.

Reply:
I mean event. At this you are aggregating events. Command is supposed to be already validated and event created. It doesn’t belong here.

Reply:
this is part where we validate events -> eg duplicit correlation id

Reply:
Where do we validate duplicit correlation id of the command? Correlation id is created by us, we do not validate if such a correlation id exist already. Only matching is done when response comes, but this  requires loading of events and finding if one matching that correlation id is there. ValidUntil is static field independent of current state. 
Your argument that it can expire during events aggregating is not well placed because same can happen with the token. Additionally, we should always require at least +100ms because event can be pushed to nats but at the moment its processed by coap gateway it might be already expired. 

As I am thinking about it, the best would be to not use unix timestamp - validUntil but ttl, defined as number of ms command is valid from counted from the moment command is accepted. Command accept response might contain value, unix timestamp when it expires. This would be much better design and you don’t need any validation if we require that it must be 0 or higher than 100ms.

Reply:
Validation is processed by snapshot by this handlers https://github.com/plgd-dev/cloud/pull/356/files#diff-c0979327025ce3f41a9ec7eb64cafdcedacb80760b1981bdbbe764b31f593616R112

I don't agree with TTL. User make a command and he is responsible when command expires. With `ttl` we must send response with `validUntil`. And maybe at that time the command is invalid.

Reply:
Command request confirmation (http / grpc response) contains valid until if ttl was valid. What usecase do you have in mind that at the time of this response the command is already expired? 

Http create command with ttl -> command validation (ttl == 0 || ttl > 100) -> create pending command event -> return eventmetadata time + ttl in confirmation response to client. 

This design is much less error prone.

Reply:
No true -> eg browser created command with ttl 10s -> and send request, the response comes after 10s, but browser considers it as canceled(because it takes 10s), with validUntil backend recognize such state pending event is not processed.

Reply:
validUntil is deterministic, ttl is floating

Reply:
ValidUntil requires time synchronization. You need to know the server time and seconds drift can for sure occur and of course, if there is drift 3 seconds, and user set validUntil to his time+ 2s command might be immediately rejected. That’s why absolute time is a bad idea. 

If clients sets ttl to 10s its up to backend to cancel it in 10s + processing. It shouldnt be anymore responsibility of the client if this query param is there.

Reply:
cloud must have synchronized time via NTP - for aws/azure they are responsible for it.

| If clients sets ttl to 10s its up to backend to cancel it in 10s + processing. It shouldnt be anymore responsibility of the client if this query param is there.

But the client doesn't want to execute the command when he cancel operation! For your scenario it same as you set validUntil=0. You don't care when the command is executed.

Reply:
Ntp synchronization is not enough. Ttl is from clients usually send as timeout in seconds exactly to prevent such an errors and to not rely on some time synchronization. Using absolute times calculated on the client side leads to errors. 

If the client set ttl and cancels earlier, it should not cancel the command. It should only cancel waiting for an response but the command is already registered.

Reply:
I can only add ttl but validUntil will stay as is.

Reply:
Why we should have both of them? They serve the same purpose.

Reply:
I still see 2 use cases:
- user wants to define absolute time (validUntil), it is from created request by user
- user wants to define drift time(ttl), validUntil will be defined by resource-aggregate

Reply:
Eg GetResourceFromDevice is a command where I want to use valid until - if I cancel by timeout command I don't want to get any more data from device.

Reply:
In your example, you can use just ttl and send cancel command - new api. Async=False query parameter we are going to introduce is about **synchronously waiting for a response**. Canceling waiting != canceling command. Command cancelation shall not be done by cancelling the session. 

regarding your user usecases, user have just one usecase. Set the time when the command becomes invalid. You described two technical solutions to do the same, not two user scenarios. 🙂

Reply:
ok so I'm for first scenario 😄

Reply:
You mean first technical solution. Me not. So let’s clarify with API consumers. I will write an email to few of them this evening. Until we have a feedback, let’s put this MR aside. 

@Danielius1922 please provide also your input. Thanks.

Reply:
> eg browser created command with ttl 10s -> and send request, the response comes after 10s, but browser considers it as canceled(because it takes 10s), with validUntil backend recognize such state pending event is not processed.

Same can happen with the validUntil. Device comes online second before expiration and command is executed. Execution takes 2 seconds. ValidUntil is by client considered as expired, but it's being processed by the device. :)

> But the client doesn't want to execute the command when he cancel operation! 
If client sent the command, this command is registered and event commandPending created. If client cancels HTTP session on which he is synchronously waiting for a response, this is not canceling the pending command. Pending command stays there.

Reply:
Fixed